mindmeld/components/custom_action.py

```diff
@@ -0,0 +1,187 @@
+
+        Args:
+            request (Request)
+            responder (DialogueResponder)
+
+        Returns:
+            (bool)
+        """
+        return await self.invoke(request, responder, async_mode=True)
+
+    def _process(self, json_data, responder):
+        status_code, result_json = self.post(json_data)
+        return self._process_post_response(status_code, result_json, responder)
+
+    async def _process_async(self, json_data, responder):
+        status_code, result_json = await self.post_async(json_data)
+        return self._process_post_response(status_code, result_json, responder)
+
+    def _process_post_response(self, status_code, result_json, responder):
+        if status_code == 200:
+            for field in RESPONSE_FIELDS:
+                if field not in result_json:
+                    logger.warning(
+                        "`%s` not in the response of custom action %s",
+                        field,
+                        self._name,
+                    )
+            if self.overwrite:
+                responder.frame = result_json.get("frame", {})
+                responder.directives = result_json.get("directives", [])
+            else:
+                responder.frame.update(result_json.get("frame", {}))
+                responder.directives.extend(result_json.get("directives", []))
+            return True
+        else:
+            logger.error(
+                "Error %s trying to reach custom action server %s.",
+                status_code,
+                self.url,
+            )
+            return False
+
+    def post(self, json_data):
+        result = requests.post(url=self.url, json=json_data)
+        if result.status_code == 200:
+            return 200, result.json()
+        else:
+            return result.status_code, {}
+
+    async def post_async(self, json_data):
+        async with aiohttp.ClientSession() as session:
+            async with session.post(self.url, json=json_data) as response:
+                if response.status == 200:
+                    return 200, await response.json()
+                else:
+                    return response.status, {}
+
+    def __repr__(self):
+        return self._name
+
+    def __str__(self):
+        return self._name
+
+
+class CustomActionSequence:
+    """
+    This class implements a sequence of custom actions
+    """
+
+    def __init__(self, actions, config, overwrite=None):
+        self.actions = [
+            CustomAction(action, config, overwrite=overwrite) for action in actions
+        ]
+
+    def invoke(self, request, responder):
+        for action in self.actions:
+            result = action.invoke(request, responder)
+            if not result:
+                logger.warning("Failed to invoke action %s.", action)
+                return False
+        return True
+
+    async def invoke_async(self, request, responder):
+        for action in self.actions:
```

Comment:
You can do this:
```
    await asyncio.gather(
        [action.invoke_async(request, responder) for action in self.actions]
    )
```

Reply:
You can't log the warning if the action fails to execute